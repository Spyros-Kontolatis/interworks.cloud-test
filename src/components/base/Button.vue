<template>
  <button class="border border-primary relative" :class="[...buttonClasses]">
    <Loader v-if="loading" class="w-8 h-8 top-3 left-2" />

    <slot></slot>
  </button>
</template>

<script setup lang="ts">
import { defineProps, computed } from "vue";
import {
  sizes,
  textColors,
  bgColors,
  rounded,
  outline,
} from "@/config/components/base/button";
import type { ButtonProps } from "@/types/components/base/button";
import Loader from "./Loader.vue";

const props = defineProps<ButtonProps>();

const buttonClasses = computed(() => {
  return [
    bgColors[props.color],
    textColors[props.color],
    sizes[props.size],
    props.rounded && rounded[props.size],
    props.outline && outline[props.color],
    (props.disabled || props.loading) &&
      "opacity-50 cursor-not-allowed pointer-events-none",
  ];
});
</script>
