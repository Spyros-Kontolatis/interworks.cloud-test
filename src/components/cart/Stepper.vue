<template>
  <div class="flex">
    <Button
      size="sm"
      color="primary"
      outline
      rounded
      :disabled="step === 1 || loading"
      class="rounded-r-none"
      @click="$emit('update:step', step - 1, 'remove')"
      ><div class="font-bold text-lg">-</div></Button
    >
    <div class="border border-primary py-2 px-4">{{ step }}</div>
    <Button
      size="sm"
      color="primary"
      rounded
      :disabled="loading"
      @click="$emit('update:step', step + 1, 'add')"
      class="rounded-l-none"
      ><div class="font-bold text-lg">+</div></Button
    >
  </div>
</template>

<script setup lang="ts">
import { StepperProps } from "@/types/components/cart/stepper";
import Button from "../base/Button.vue";
import { defineProps, defineEmits } from "vue";

defineProps<StepperProps>();
defineEmits(["update:step"]);
</script>
