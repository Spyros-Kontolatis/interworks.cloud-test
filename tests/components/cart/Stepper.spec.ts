import { shallowMount } from "@vue/test-utils";
import Stepper from "../../../src/components/cart/Stepper.vue";
import Button from "@/components/base/Button.vue";

describe("Stepper.vue", () => {
  it("renders the component with required props", () => {
    const wrapper = shallowMount(Stepper, {
      props: {
        step: 1,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("successfully increments/decrements stepper", async () => {
    const wrapper = shallowMount(Stepper, {
      props: {
        step: 1,
        "onUpdate:step": (newStep: number, _action: "remove" | "add") =>
          wrapper.setProps({ step: newStep }),
      },
    });
    expect(wrapper.element).toMatchSnapshot();
    const [decrementBtn, incrementBtn] = wrapper.findAllComponents(Button);
    incrementBtn.vm.$emit("click");
    await wrapper.vm.$nextTick();
    expect((wrapper.props() as { step: number }).step).toBe(2);
    decrementBtn.vm.$emit("click");
    await wrapper.vm.$nextTick();
    expect((wrapper.props() as { step: number }).step).toBe(1);
  });
});
