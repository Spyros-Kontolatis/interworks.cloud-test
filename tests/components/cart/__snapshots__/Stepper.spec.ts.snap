// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper.vue renders the component with required props 1`] = `
<div
  class="flex"
>
  <button-stub
    class="rounded-r-none"
    color="primary"
    disabled="true"
    loading="false"
    outline="true"
    rounded="true"
    size="sm"
  />
  <div
    class="border border-primary py-2 px-4"
  >
    1
  </div>
  <button-stub
    class="rounded-l-none"
    color="primary"
    disabled="false"
    loading="false"
    outline="false"
    rounded="true"
    size="sm"
  />
</div>
`;

exports[`Stepper.vue successfully increments/decrements stepper 1`] = `
<div
  class="flex"
>
  <button-stub
    class="rounded-r-none"
    color="primary"
    disabled="true"
    loading="false"
    outline="true"
    rounded="true"
    size="sm"
  />
  <div
    class="border border-primary py-2 px-4"
  >
    1
  </div>
  <button-stub
    class="rounded-l-none"
    color="primary"
    disabled="false"
    loading="false"
    outline="false"
    rounded="true"
    size="sm"
  />
</div>
`;
